ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java

```diff
@@ -0,0 +1,573 @@
+  private final String clusterName;
+  private final MetricRegistry metricRegistry;
+  private final ClusterMapConfig clusterMapConfig;
+  private final Map<String, DcZkInfo> dcToDcZkInfo = new HashMap<>();
+  private final Map<String, AmbryPartition> partitionNameToPartitionIdMap = new ConcurrentHashMap<>();
+  private final Map<String, AmbryDataNode> instanceNameToDataNodeIdMap = new ConcurrentHashMap<>();
+  private final Map<AmbryPartition, Set<AmbryReplica>> partitionIdToReplicaIds = new ConcurrentHashMap<>();
+  private final Map<AmbryDataNode, Set<AmbryReplica>> dataNodeIdToReplicaIds = new ConcurrentHashMap<>();
+  private final Map<AmbryDataNode, Set<AmbryDisk>> dataNodeIdToDiskIds = new ConcurrentHashMap<>();
+  private final Map<ByteBuffer, AmbryPartition> partitionMap = new ConcurrentHashMap<>();
+  private final HelixClusterManagerCallback helixClusterManagerCallback;
+  private long clusterWideRawCapacityBytes;
+  private long clusterWideAllocatedRawCapacityBytes;
+  private long clusterWiseAllocatedUsableCapacityBytes;
+  final HelixClusterManagerMetrics helixClusterManagerMetrics;
+
+  /**
+   * Instantiate a HelixClusterManager.
+   * @param clusterMapConfig the {@link ClusterMapConfig} associated with this manager.
+   * @param instanceName the String representation of the instance associated with this manager.
+   * @throws IOException if there is an error in parsing the clusterMapConfig or in connecting with the associated
+   *                     remote Zookeeper services.
+   */
+  HelixClusterManager(ClusterMapConfig clusterMapConfig, String instanceName) throws IOException {
+    this.clusterMapConfig = clusterMapConfig;
+    clusterName = clusterMapConfig.clusterMapClusterName;
+    metricRegistry = new MetricRegistry();
+    helixClusterManagerCallback = new HelixClusterManagerCallback();
+    helixClusterManagerMetrics = new HelixClusterManagerMetrics(metricRegistry, helixClusterManagerCallback);
+    try {
+      Map<String, String> dataCenterToZkAddress =
+          parseZkJsonAndPopulateZkInfo(clusterMapConfig.clusterMapDcsZkConnectStrings);
+      for (Map.Entry<String, String> entry : dataCenterToZkAddress.entrySet()) {
+        String dcName = entry.getKey();
+        String zkConnectStr = entry.getValue();
+        HelixManager manager =
+            HelixManagerFactory.getZKHelixManager(clusterName, instanceName, InstanceType.SPECTATOR, zkConnectStr);
+        manager.connect();
+        ClusterChangeListener clusterChangeListener = new ClusterChangeListener();
+        DcZkInfo dcZkInfo = new DcZkInfo(dcName, zkConnectStr, manager, clusterChangeListener);
+        dcToDcZkInfo.put(dcName, dcZkInfo);
+      }
+      // Now initialize by pulling information from Helix.
+      initialize();
+      // Now register listeners to get notified on change.
+      for (DcZkInfo dcZkInfo : dcToDcZkInfo.values()) {
+        dcZkInfo.helixManager.addExternalViewChangeListener(dcZkInfo.clusterChangeListener);
```

Comment:
We are not currently using it, but this notification will come once per resource when nodes go down as I understand it, and it will contain the partitions and their instance mappings and so on. We will make use of this in the future.